### PR TITLE
New event: "ZWave.incomingPacket" (srcNodeId, dstNodeId, payloadDec)

### PR DESCRIPTION
**Pull request by PeteH32**
New event so modules can get raw incoming Z-Wave packets, after decapsulation and decryption.

Here is example of how a module could subscribe to this event, and process it's arguments:

```        // Setup listener for incoming Z-Wave packets
        function decToHex(decimal) {
                return ("00" + decimal.toString(16).toUpperCase()).slice(-2);
        }

        this.incomingPacketHandler = function(srcNodeId, dstNodeId, payloadDec) {
                var payloadHex = [];
                for (var i = 0; i < payloadDec.length; i++) {
                        payloadHex += decToHex(payloadDec[i]) + " ";
                }
                console.log("LOG: incomingPacketHandler(): srcNodeId = " + srcNodeId + ", dstNodeId = " + dstNodeId + ", payloadHex = [ " + payloadHex + "]");
        };

        this.controller.on("ZWave.incomingPacket", this.incomingPacketHandler);
```

### CODE REVIEW

**Comment by PoltoS** [comment]
Nice idea, but I would make it disabled by default not to load the CPU with huge emit code for every packet. Worth to check the impact of this change on a big network with many devices.

But as a feature this is interesting. BTW, what do you do with this raw command?

**Comment by pete-heirendt-vacasa** [comment]
Thanks for the feedback!  I'll push a change to have it be disabled by default.

The primary reason we are passing raw commands to our module is so we can more easily process notifications and reports from different devices. If the module developer is already familiar with the Z-WaVE specification for the below command classes, it is very easy to work with the raw hex codes directly.

- **COMMAND_CLASS_NOTIFICATION** - Lots of commands in this class!  :-)
- **COMMAND_CLASS_DOOR_LOCK_LOGGING** - RECORD_REPORT
- **COMMAND_CLASS_DOOR_LOCK** - DOOR_LOCK_OPERATION_REPORT
- **COMMAND_CLASS_SCHEDULE_ENTRY_LOCK** - multiple *_REPORT commands
- **COMMAND_CLASS_TIME_PARAMETERS** - TIME_PARAMETERS_REPORT

To summarize, it's mainly to process the large variety of reports and notifications, assuming the module developer is familiar with the Z-Wave specification for the above command classes.

**Comment by PoltoS** [comment]
This is a vicious practice ;)

Z-Way has a very nice parser of Notification, Door Lock Logging, Door Lock and Schedule Entry Lock 
 CCs. Time Parameters is handled intrinsically - nothing to expose to the user.

**Comment by PeteH32** [comment]
> This is a vicious practice ;)
> 
> Z-Way has a very nice parser of Notification, Door Lock Logging, Door Lock and Schedule Entry Lock CCs.

That is a good point!  The parsers you mention are very rich and thorough, and we use the Expert UI's and the "Expert commands" a lot (they are awesome!).

But I ran into one problem (I think) for our specific use-case -- our module/app needs to capture **every** notification/report from the device, even when the device sends a Notification command on it's own (not in response to a GET from the controller). We also need to handle case where the device sends several notifications in a row. 

When we tried using the Z-Way Data Model, for example using this data object (I might not have syntax correct - this is just from memory):

- `devices[n].instances[m].Alarm.data[0x06].eventParameters`

However, that appears to only have the most recent event, and if the device sent 3 notifications in a row, we could miss the first 2 notifications. At least that was our guess, and that is why we started going the more vicious approach.  By creating the new "ZWave.incomingPacket" event, the module/app will get **every** notification sent by the device.

**Comment by PoltoS** [comment]
Looks you missed the Z-Way concept of doing things.

`devices[n].instances[m].Alarm.data[0x06].bind(function(data) { console.logJS(data); })` will capture every event and allow you to access it's data.

**Comment by PoltoS** [comment]
The action is executed on the server side. You can do some actions based on those events.
